### PR TITLE
Add node-mssql-sqli rule

### PR DESCRIPTION
**Pull request by inkz**
new rule for SQL injection via `mssql` module

### CODE REVIEW

**Comment by inkz** [comment]
failure related to a different rule

File: javascript/lang/security/audit/sqli/node-mssql-sqli.yaml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            $QUERY += $X
+            ...
+        - pattern-inside: |
+            $QUERY = $X.concat($Y)
+            ...
+        - pattern-inside: |
+            $QUERY = `...${...}...`
+            ...
+      - pattern-not-inside: |
+          $QUERY += "..."
```
Comment (clintgibler) [inline javascript/lang/security/audit/sqli/node-mssql-sqli.yaml:46]: With the `+=`, couldn't user input still be in `$QUERY`? Something like:

```javascript
var foo = "select * from [foobar] WHERE user_id =" + userInput);
foo += " sort desc";
pool.query(foo);
```

Maybe it's good to keep and there are cases I'm not thinking about properly, just wanted to ask. Sould it be `$QUERY = "..."`?

Would `metavariable-pattern` potentially help simplify?

**Review by clintgibler — APPROVED** [review]
I had a small question and I wonder if any cleaning up/improving could be done, but overall lgtm 👍

File: javascript/lang/security/audit/sqli/node-mssql-sqli.yaml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+          ...
+      - pattern-not-inside: |
+          $QUERY = "..." + "..."
```
Comment (emjin) [inline javascript/lang/security/audit/sqli/node-mssql-sqli.yaml:49]: @IagoAbal This seems like a feasible constant prop improvement.

Comment (emjin) [reply javascript/lang/security/audit/sqli/node-mssql-sqli.yaml:49]: (If not there already)

File: javascript/lang/security/audit/sqli/node-mssql-sqli.yaml (hunk 1 above)
Comment (emjin) [reply javascript/lang/security/audit/sqli/node-mssql-sqli.yaml:46]: Not sure if metavariable-pattern would be great here. I think this seems like something taint mode could do?

**Comment by minusworld** [comment]
The failing `test-develop` check is expected until https://github.com/returntocorp/semgrep-rules/pull/1427 is merged, which requires the next release of Semgrep.

**Comment by inkz** [comment]
@clintgibler @emjin very good questions, all SQLi rules look like a perfect candidates for taint mode, but some further investigation needed for this. I decided to push this rule first to unblock our user (here: https://github.com/returntocorp/semgrep-rules/issues/1430)

I am going to vote for Sql Injection be our next problem-centric / cheat sheet series :+1:

File: javascript/lang/security/audit/sqli/node-mssql-sqli.yaml (hunk 2 above)
Comment (IagoAbal) [reply javascript/lang/security/audit/sqli/node-mssql-sqli.yaml:49]: You shouldn't need `"..." + "..."` with const-prop enabled, just `"..."` should work.

**Comment by JasonCubic** [comment]
@inkz 

I just tested this out, and it works so great!  :-)

One small note though.

The parameter code example on the description for this rule isn't correct.  using parameters in mssql is a little more complex than the example in the rule description indicates.  https://www.npmjs.com/package/mssql#input-name-type-value

I would suggest either leaving the example out or linking to the mssql docs.  because this:
```
const dbResult = await request.query('SELECT * FROM [ramshackle].[dbo].[users] WHERE user_id = $1', [id]);
```
only returns the user_id of 1 no matter what the id variable is set to (it ignores the $).

I think mssql makes you do something like this:
```
request.input('USER_ID', mssql.Int, id);
const dbResult = await request.query('SELECT * FROM [ramshackle].[dbo].[users] WHERE user_id = @USER_ID');
```

**Comment by inkz** [comment]
@JasonCubic thanks a lot for help! :+1:
I used `postgresql` rule as a template and forgot to change description.
Created a fix: https://github.com/returntocorp/semgrep-rules/pull/1449

**Pull request merged**